test_runner/regress/test_remote_storage.py

```diff
@@ -784,91 +782,86 @@ def test_compaction_waits_for_upload(
+        path = env.pageserver.timeline_dir(tenant_id, timeline_id) / name
+        assert path.exists(), "while uploads are stuck the layers should be present on disk"
 
-    checkpoint_thread = threading.Thread(target=checkpoint_in_background)
-    checkpoint_thread.start()
+    # now this will do the L0 => L1 compaction and want to remove our layers
+    client.timeline_checkpoint(tenant_id, timeline_id)
 
-    compact_thread = threading.Thread(target=compact_in_background)
-    compact_thread.start()
+    # as uploads are paused, the L0 layers should still be with us
+    for name in upload_stuck_layers:
+        path = env.pageserver.timeline_dir(tenant_id, timeline_id) / name
+        assert path.exists(), "uploads are stuck still over compaction"
 
-    try:
-        # Start the checkpoint, see that it blocks
-        log.info("Waiting to see checkpoint hang...")
-        time.sleep(5)
-        assert checkpoint_result.empty()
-
-        # Start the compaction, see that it finds work to do but blocks
-        compact_barrier.wait()
-        log.info("Waiting to see compaction hang...")
-        time.sleep(5)
-        assert compact_result.empty()
-
-        # This is logged once compaction is started, but before we wait for operations to complete
-        assert env.pageserver.log_contains("compact_level0_phase1 stats available.")
-
-        # Once we unblock uploads the compaction should complete successfully
-        log.info("Disabling failpoint")
-        client.configure_failpoints(("before-upload-layer-pausable", "off"))
-        log.info("Awaiting compaction result")
-        assert compact_result.get(timeout=10) is None
-        log.info("Awaiting checkpoint result")
-        assert checkpoint_result.get(timeout=10) is None
-
-    except Exception:
-        # Log the actual failure's backtrace here, before we proceed to join threads
-        log.exception("Failure, cleaning up...")
-        raise
-    finally:
-        compact_barrier.abort()
+    compacted_layers = client.layer_map_info(tenant_id, timeline_id).historic_by_name()
+
+    overlap = compacted_layers.intersection(upload_stuck_layers)
+    assert len(overlap) == 0, "none of the L0's should remain after L0 => L1 compaction"
+
+    def gcs_completed():
+        m = client.get_metric_value("pageserver_layer_gcs_count_total", {"state": "completed"})
+        if m is None:
+            return 0
+        return int(m)
 
-        checkpoint_thread.join()
-        compact_thread.join()
+    # if initdb created an initial delta layer, it might already be gc'd
+    assert gcs_completed() <= 1
+
+    client.configure_failpoints(("before-upload-layer-pausable", "off"))
 
     # Ensure that this actually terminates
     wait_upload_queue_empty(client, tenant_id, timeline_id)
 
+    def until_gcs_completed():
+        gcs = gcs_completed()
+        log.info(f"gcs: {gcs}")
+        assert gcs >= len(upload_stuck_layers)
+
+    wait_until(10, 1, until_gcs_completed)
```

Comment:
I don't see us do any garbage collection here, which I assume `gcs` means (?)

Reply:
I remember objecting to calling the "dangling file deletion" "garbage collection" :/

Reply:
So the span would be "layer_(un|de)dangle"? :) Perhaps instead `layer_delete` as that is what we are calling it over at RemoteTimelineClient.

Reply:
This continues on a follow-up I had code ready: #5902.